### PR TITLE
Adapter debug support

### PR DESCRIPTION
**Pull request by turquoiseowl**
Thank you for this useful project.

The one thing missing for me was script debugging from the Test Adapter. The goal was to set a breakpoint in a test script, right click the entry for the test in Test Explorer, select Debug Selected Test, then have the debugger load and break at that BP.

So I've had a go at getting that to work. It's quite a large PR so no worries if it doesn't get merged, but wanted to communicate how you might go about implementing that feature.

Essentially, I've taken a look at how Visual Studio debugs a script and worked out how to do that programatically when the test adapter is executed in IsBeingDebugged==true mode.

The test result are communicated in the browser, as with the OpenInBrowser mode.

### CODE REVIEW

File: Chutzpah/Chutzpah.csproj [hunk 1]
```diff
@@ -40,6 +40,12 @@
     <Reference Include="AntiXssLibrary">
       <HintPath>..\packages\AntiXSS.4.2.1\lib\net40\AntiXssLibrary.dll</HintPath>
     </Reference>
+    <Reference Include="EnvDTE, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
```
Comment (mmanela) [inline Chutzpah/Chutzpah.csproj]: Having Chutzpah core library take a reference on envDTE seems bad. This means chutzpah.exe will have a dependency on a VS install.

**Comment by turquoiseowl** [comment]
Yes, I agree.

I considered doing a sub class of ProcessHelper in VS2012.TestAdapter and DI that into TestRunner, but then thought the functionality might also be wanted from VisualStudioContextMenu.

**Comment by turquoiseowl** [comment]
Ahh, VS.Common would be the place. I still need help on injecting into TestRunner.

**Comment by mmanela** [comment]
Wow! This is amazing. I took a quick pass through the code and it is very impressive. I have not tested out but assuming this works it is pretty cool. I have looked into doing this a couple times but never had the time to really figure it all out.

A few questions/comments
1. How long did this take you to do? 

2. As mentioned above the forcing DTE reference in the chutzpah core dll might not work. As you mentioned we could put in VS common but I could also see value in someone from the command line writing /launchInVSDebugger. But I don't think that scenario is as important but if we wanted it maybe there is a way to delay bind against DTE so we don't need a direct reference on it? If that is not possible then maybe just moving to VS.Common is enough. And yes it should work in both VS plugins.

3.  Will this work to let you natively debug files like TypeScript or CoffeeScript if you have the sourcemaps present? Or is it only for JS files?

**Comment by turquoiseowl** [comment]
Thanks for the positive feedback. If I was being paid, it took a week, but as I'm not it took last night to work it out and today to code it up properly.

2. I can move the DTE reference into VS.Common. Only thing is I may need to alter some of the DI stuff for TestRunner to get a sub class of ProcessHelper injected. I'm just a bit nervous about making too many changes to the core because the unit tests hang about 3/4 way through. The alternative would be to add a setter like ```TestRunner.SetProcessHelper``` but that smells bit. Any thoughts?

3. What ever script debugging works from Visual Studio should work here. From googling it looks like that means typescript will work. Not sure about coffeescript. The manual way to go about it is ```VS | Debug | Attach to process``` and the ```Attach To``` setting which needs to be ```Script```. Then attach to a running iexplore.exe instance and you 'should' find BPs set in your script get hit.

**Comment by mmanela** [comment]
One idea is to pass the your code that knows how to attach to a process as an object in TestOptions. Maybe we generalize that and make a TestLaunchMode Context object. So you can pass TestLaunch: Headless, Browser, Custom. And then for custom you must pass a ITestLauncher object as part of the TestOptions. 

What do you think of that?

**Comment by turquoiseowl** [comment]
Yes, good idea. I'll rework in the morning unless I hear further.

**Comment by turquoiseowl** [comment]
When I add a new class to the VS.Common project -- even just empty class -- the test adapter no longer runs. I must be missing something obvious. Any ideas?

**Comment by mmanela** [comment]
What is the error? You can check the ActivityLog in VS for details (or debug).

**Comment by turquoiseowl** [comment]
Strange, there's no error:

```
  <entry>
    <record>227</record>
    <time>2015/07/22 13:39:23.965</time>
    <type>Information</type>
    <source>Microsoft.VisualStudio.CommonIDE.ExtensibilityHosting.VsShellComponentModelHost</source>
    <description>Successfully loaded component assembly from cache</description>
    <path>C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO 11.0\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\TESTWINDOW\EXTENSIONS\Chutzpah.VS.Common.dll</path>
  </entry>
  <entry>
    <record>228</record>
    <time>2015/07/22 13:39:23.965</time>
    <type>Information</type>
    <source>Microsoft.VisualStudio.CommonIDE.ExtensibilityHosting.VsShellComponentModelHost</source>
    <description>Successfully loaded component assembly from cache</description>
    <path>C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO 11.0\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\TESTWINDOW\EXTENSIONS\Chutzpah.VS2012.TestAdapter.dll</path>
  </entry>
```
I'm wondering whether the cache mentioned has something to do with it. No problem when adding empty class to Chutzpah.VS2012.TestAdapter.dll.

**Comment by mmanela** [comment]
Maybe blow away the extension cache?

-----Original Message-----

**Pull request closed without merging**